embedded-tests/src/test/java/org/apache/druid/testing/embedded/query/QueryVirtualStorageTest.java

```diff
@@ -152,88 +158,141 @@ void testQueryPartials()
         "select count(*) from \"%s\" WHERE __time >= TIMESTAMP '2015-09-12 14:00:00' and __time < TIMESTAMP '2015-09-12 19:00:00'",
         "select count(*) from \"%s\" WHERE __time >= TIMESTAMP '2015-09-12 19:00:00' and __time < TIMESTAMP '2015-09-13 00:00:00'"
     };
-    final long[] expectedResults = new long[] {
-        9770,
-        10524,
-        10267,
-        8683
-    };
+    final long[] expectedResults = new long[]{9770, 10524, 10267, 8683};
+    final long[] expectedLoads = new long[]{8L, 6L, 5L, 5L};
+
+
+    LatchableEmitter emitter = historical.latchableEmitter();
+    // sleep to clear out the pipe to get zerod out storage monitor metrics and then flush (which clears out the
+    // internal events stores in test emitter)
+    sleepForStorageMonitor();
```

Comment:
Can we avoid the sleep somehow? Maybe by waiting for some specific metric to be emitted?

Reply:
i had some trouble getting waiting for events from the same monitor to work, looking into the code i think i see why, waitForEvent doesn't necessarily wait for the next event, it finds any processed matching event too, i believe i misunderstood what this method did, I think i need a version that skips any already emitted events and only waits for the next one

Reply:
> I think i need a version that skips any already emitted events and only waits for the next one

I see. That seems like a reasonable use case. I think you can add a method `waitForNextEvent` but we would need to call out that it would be susceptible to race conditions and may miss out on events and cause the test to wait forever.

Reply:
this is more or less exactly what i have done, just haven't got around to pushing up a commit yet tidying some other stuff up